File: algorithm/go/2131-longestPalindromeByConcatenationTwoLetterWords.go
//go:build ignore

package main

import "fmt"

func longestPalindrome(words []string) int {
    word_cnt := make(map[string]int)

	for _, word := range words {
		word_cnt[word]++
	}

	res := 0
	var center int

	for word, cnt := range word_cnt {
		if word[0] == word[1] {
			if center == 0 {
				center = cnt % 2
			}
			res += (cnt - cnt % 2) * 2
		} else {
			rev := string([]byte{word[1], word[0]})
			if rev > word {
				res += min(cnt, word_cnt[rev]) * 4
			}
		}
	}

	return res + center * 2
}

func main() {
	words := []string{"ab","ty","yt","lc","cl","ab"}
	fmt.Print(longestPalindrome(words))
}

File: algorithm/python/2131-longestPalindromeByConcatenationTwoLetterWords.py
from typing import List
from collections import defaultdict

class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        word_dictionary = defaultdict(int)
        for word in words:
            word_dictionary[word] += 1
        
        res = 0
        check = False
        for word in list(word_dictionary.keys()):
            if word[0] == word[1]:
                if word_dictionary[word] % 2:
                    check = True
                res += (word_dictionary[word] - (word_dictionary[word] % 2))
            else:
                palin = word[1] + word[0]
                res += min(word_dictionary[word], word_dictionary.get(palin, 0)) 
            
        return (res + check) * 2
    
def main():
    sol = Solution()
    print(sol.longestPalindrome(["lc", "cl", "gg"]))

if __name__ == "__main__":
    main()
    